tests/returntypes/test_new_qnspsa.py

```diff
@@ -0,0 +1,417 @@
+        (on/off) are tested.
+        """
+        regularization = 1e-3
+        stepsize = 1e-2
+        opt_blocking = qml.QNSPSAOptimizer(
+            stepsize=stepsize,
+            regularization=regularization,
+            finite_diff_step=finite_diff_step,
+            resamplings=1,
+            blocking=True,
+            history_length=5,
+            seed=seed,
+        )
+        opt_no_blocking = deepcopy(opt_blocking)
+        opt_no_blocking.blocking = False
+        # target opt is used to reproduce the result with step()
+        target_opt = deepcopy(opt_blocking)
+
+        qnode, params_shape = get_single_input_qnode()
+        params = np.random.rand(*params_shape)
+
+        new_params_blocking_res, qnode_blocking_res = opt_blocking.step_and_cost(qnode, params)
+        with pytest.warns(UserWarning):
+            new_params_expected = target_opt.step(qnode, params)
+        # analytical expression of the qnode
+        qnode_expected = np.cos(params[0][0] / 2) ** 2 - np.sin(params[0][0] / 2) ** 2 * np.cos(
+            params[0][1]
+        )
+        assert np.allclose(new_params_blocking_res, new_params_expected)
+        assert np.allclose(qnode_blocking_res, qnode_expected)
+
+        new_params_no_blocking_res, qnode_no_blocking_res = opt_no_blocking.step_and_cost(
+            qnode, params
+        )
+        assert np.allclose(new_params_no_blocking_res, new_params_expected)
+        assert np.allclose(qnode_no_blocking_res, qnode_expected)
+
+    def test_step_and_cost_from_multi_input(self, finite_diff_step, seed):
+        """Test step_and_cost() function with the multi-input qnode."""
+        regularization = 1e-3
+        stepsize = 1e-2
+        opt = qml.QNSPSAOptimizer(
+            stepsize=stepsize,
+            regularization=regularization,
+            finite_diff_step=finite_diff_step,
+            resamplings=1,
+            blocking=True,
+            history_length=5,
+            seed=seed,
+        )
+        # target opt is used to reproduce the random sampling result
+        target_opt = deepcopy(opt)
+
+        qnode = get_multi_input_qnode()
+        params = [np.array(1.0) for _ in range(2)]
+        # this single-step result will be different from the one from the single-input qnode, due to the
+        # different order in sampling perturbation directions.
+        new_params_res, qnode_res = opt.step_and_cost(qnode, *params)
+
+        # test the expectation value
+        qnode_expected = np.cos(params[0] / 2) ** 2 - np.sin(params[0] / 2) ** 2 * np.cos(params[1])
+        assert np.allclose(qnode_res, qnode_expected)
+
+        # test the next-step parameter
+        _, grad_dirs = target_opt._get_spsa_grad_tapes(qnode, params, {})
+        _, tensor_dirs = target_opt._get_tensor_tapes(qnode, params, {})
+        qnode_finite_diff = get_grad_finite_diff(params, finite_diff_step, grad_dirs).reshape(1, 1)
```

Comment:
How come a new shape was required here?

Reply:
This test was copied from the original test file. My understanding is that the output of `get_grad_finite_diff` is a scalar, which needs to be reshaped to broadcast correctly with `grad_dirs`.

Reply:
This test was copied from the original test file, but indeed it looks like the extra reshape is unnecessary.